torch/distributed/_spmd/api.py

```diff
@@ -219,8 +221,117 @@ def f_true():
+
+    # 2. Override target submodules (e.g., MoE) with dummy replacements
+    if module_override:
+        accessor = NamedMemberAccessor(mod)
+
+        for typ, override in module_override.items():
+            for name, submodule in mod.named_modules():
+                if isinstance(submodule, typ):
+                    accessor.swap_submodule(
+                        name, override.replacement(submodule)
+                    )
+
+    # 3. Trace statelss version of the train_step
+    params_and_buffers: Dict[str, Union[torch.Tensor, nn.Parameter]] = {
+        **dict(mod.named_parameters(remove_duplicate=False)),
+        **dict(mod.named_buffers(remove_duplicate=False)),
+    }
+
+    named_states = {}
+    if opt is not None:
+        opt_states, spec = pytree.tree_flatten(dict(opt.state))
+
+        # Pass named_states instead of opt.state to stateless_func, because
+        # the later uses nn.Parameter as key. During tracing, we need to
+        # make sure optimizers can find the states using proxy tensors.
+        for n, p in params_and_buffers.items():
+            if p in opt.state:
+                # opt.state's key type is string, but optimizer uses
+                # Parameter as keys
+                named_states[n] = opt.state[p]  # type: ignore[index]
+
+    # Lift states and parameters as function arguments so that make_fx
+    # can trace operations applied to them.
+    def stateless_func(
+        func, args, kwargs, named_states, params_and_buffers
+    ):
+        with stateless._reparametrize_module(
+            cast(nn.Module, mod), params_and_buffers
+        ), _rematerialize_optimizer(
+            opt, named_states, params_and_buffers
+        ) if opt else nullcontext():
+            ret = func(*args, **kwargs)
+            # make sure updated parameters are returned
+            return ret, list(mod.parameters())  # type: ignore[union-attr]
+
+    # FIXME: Using symbolic tracing to work around. Otherwise it hits
+    # shape mismatch error, as we use local inputs to trace local graph
+    # and use DTensor to expand operators, where DTensor's shape is the
+    # global shape.
+    with _enable_compile():
+        # FIXME: functionalize crashes with
+        # "UnsupportedFakeTensorException: meta converter nyi"
+        gm = make_fx(
+            partial(stateless_func, func),
+            tracing_mode="symbolic",
+            _allow_non_fake_inputs=True,
+        )(args, kwargs, named_states, params_and_buffers)
+
+    # 4. Use DTensor to insert collectives
+    gm, name_to_spec = _dtensor_expand(
+        gm, args, kwargs, named_states, params_and_buffers
+    )
+
+    # 5. Replace previously inserted dummy ones with real graphs.
+    if module_override:
+        for _, override in module_override.items():
+            gm = override.transform(gm, name_to_spec)
+
+    return _CompiledResult(gm, mod, opt, named_states, params_and_buffers)
+
+
+COMPILED_OBJECT_KEY = "_compiled_obj"
```

Comment:
this can simply be an `COMPILED_OBJECT_KEY =object()` to avoid potential string name conflicts?

Reply:
I remember there was a serialization issue due to using `object()` for composable APIs. But I guess this is not a problem unless we want to serialize the `train_step`. Will change it to `object()`

Reply:
Hmm, I just remember that the convention of `__dict__` is that keys must be strings. So I will keep str but add a TODO to ensure the key is unique.